File: seminar-03/sorting_algorithms/merge_sort/java/MergeSortOptimized.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class MergeSortOptimized {

    static void sort(int[] a, int[] tmp, int l, int r) {
        if (r - l < 2) return;

        int m = (l + r) / 2;
        sort(a, tmp, l, m);
        sort(a, tmp, m, r);

        int i = l, j = m, k = l;
        while (i < m || j < r) {
            if (i == m) {
                tmp[k++] = a[j++];
            } else if (j == r) {
                tmp[k++] = a[i++];
            } else if (a[i] <= a[j]) { // <= для устойчивости
                tmp[k++] = a[i++];
            } else {
                tmp[k++] = a[j++];
            }
        }

        for (int t = l; t < r; t++) {
            a[t] = tmp[t];
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(in.readLine());
        int[] a = new int[n];

        if (n > 0) {
            String[] parts = in.readLine().split(" ");
            for (int i = 0; i < n; i++) {
                a[i] = Integer.parseInt(parts[i]);
            }
        }

        int[] tmp = new int[n];
        sort(a, tmp, 0, n);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i > 0) sb.append(' ');
            sb.append(a[i]);
        }
        System.out.println(sb);
    }
}

File: seminar-03/sorting_algorithms/merge_sort/python/merge_sort_optimized.py
def merge_sort(a):
    n = len(a)
    buf = [0] * n

    def sort(l, r):
        if r - l < 2:
            return
        m = (l + r) // 2
        sort(l, m)
        sort(m, r)

        i, j, k = l, m, l
        while i < m or j < r:
            if i == m:
                buf[k] = a[j]; j += 1
            elif j == r:
                buf[k] = a[i]; i += 1
            elif a[i] <= a[j]:  # <= для устойчивости
                buf[k] = a[i]; i += 1
            else:
                buf[k] = a[j]; j += 1
            k += 1

        for t in range(l, r):
            a[t] = buf[t]

    sort(0, n)


n = int(input())
a = list(map(int, input().split())) if n > 0 else []
merge_sort(a)
print(*a)
